lib/dynamoid/adapter.rb

```diff
@@ -5,27 +5,48 @@ module Dynamoid
   # 1) For the rest of Dynamoid, the gateway to DynamoDB.
   # 2) Allows switching `config.adapter` to ease development of a new adapter.
   # 3) Caches the list of tables Dynamoid knows about.
-  module Adapter
-    def self.tables
-      @tables
+  class Adapter
+    # MUTEX RULE:
+    # `@tables_mutex` must always be acquired before `@adapter_mutex`!
+    # This is mandated by how `#tables` works: while holding `@tables_mutex`, it
+    # needs `#adapter` to list tables.
+
+    def initialize
+      @adapter_mutex = Mutex.new
+      @tables_mutex = Mutex.new
     end
 
-    # The actual adapter currently in use: presently AwsSdk.
-    #
-    # @since 0.2.0
-    def self.adapter
-      reconnect! unless @adapter
-      @adapter
+    def tables
```

Comment:
So the double-check locking idiom is (mostly) not safe. In very rare circumstances, I think this code can fail. See: http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html